lightningnetwork/lnd
# sqldb: new package supporting sql backends

This is the second PR for https://github.com/lightningnetwork/lnd/issues/6288 

Only the latest commit applies to this PR

The [previous PR](https://github.com/lightningnetwork/lnd/pull/7354) we added the sql schema and queries for invoices.

In this PR provide the scaffolding for using the new sql stores.

The new interfaces, structs and methods are in sync with other projects like [Taproot Assets](https://github.com/lightninglabs/taproot-assets/tree/main/tapdb).
    
- **Transactional Queries**: the sqldb package defines the interfaces required to execute transactional queries to our storage interface.
    
- **Embed Migration Files**: the migration files are embedded into the binary. 
    
- **Database Migrations**: I kept the use of 'golang-migrate' to ensure our codebase remains in sync with the other projects, but it can be changed. (@joostjager provided some feedback about golang-migrate pitfalls and how [sql-migrate](github.com/rubenv/sql-migrate) can be a good subtitute)
    
- Build Flags for Conditional DB Target: flexibility to specify our database target at compile-time based on the build flags in the same way we do with our kv stores.
    
- SQL utils: generic functions used for marshal/unmarshal db data. <= I will fix the linter later
    
- Update modules: ran `go mod tidy`.

**NOTE**: the structure of this PR has changed so some of the old comments don't apply or apply to other PRs.
